src/dotnet-interactive/CodeExpansionConfiguration.cs

```diff
@@ -0,0 +1,104 @@
+{
+    private readonly IJupyterKernelSpecModule _kernelSpecModule;
+    private readonly Dictionary<string, CodeExpansion> _codeExpansions = new();
+    private bool _checkedKernelSpecs;
+
+    public CodeExpansionConfiguration(
+        IEnumerable<CodeExpansion> codeExpansions = null,
+        IJupyterKernelSpecModule kernelSpecModule = null)
+    {
+        _kernelSpecModule = kernelSpecModule;
+        if (codeExpansions is not null)
+        {
+            foreach (var codeExpansion in codeExpansions)
+            {
+                _codeExpansions.Add(codeExpansion.Info.Name, codeExpansion);
+            }
+        }
+    }
+
+    public Func<RecentConnectionList> GetRecentConnections { get; init; }
+
+    public Action<RecentConnectionList> SaveRecentConnections { get; init; }
+
+    public void AddCodeExpansion(CodeExpansion codeExpansion)
+    {
+        _codeExpansions.Add(codeExpansion.Info.Name, codeExpansion);
+    }
+
+    public async Task<CodeExpansion> GetCodeExpansionAsync(string name)
+    {
+        await EnsureKernelSpecsHaveBeenCheckedAsync();
+
+        var recentConnectionExpansions = GetExpansionsForRecentConnections();
+
+        if (recentConnectionExpansions.FirstOrDefault(e => e.Info.Name.Equals(name,StringComparison.OrdinalIgnoreCase)) is {} expansion)
+        {
+            return expansion;
+        }
+
+        return _codeExpansions.GetValueOrDefault(name);
+    }
+
+    public async Task<List<CodeExpansionInfo>> GetCodeExpansionInfosAsync()
+    {
+        await EnsureKernelSpecsHaveBeenCheckedAsync();
+
+        var expansions = GetExpansionsForRecentConnections();
+        expansions.AddRange(_codeExpansions.Values);
+        return expansions.Select(e => e.Info).ToList();
+    }
+
+    private List<CodeExpansion> GetExpansionsForRecentConnections()
+    {
+        List<CodeExpansion> infos = new();
+
+        if (GetRecentConnections is not null)
+        {
+            var recentConnectionList = GetRecentConnections();
+            infos.AddRange(recentConnectionList);
+        }
+
+        return infos;
+    }
+
+    private async Task EnsureKernelSpecsHaveBeenCheckedAsync()
+    {
+        if (_checkedKernelSpecs || _kernelSpecModule is null)
+        {
+            return;
+        }
+
+        var kernelSpecs = await _kernelSpecModule.ListKernelsAsync();
+
+        foreach (var kernelSpec in kernelSpecs)
+        {
+            var codeExpansionInfo = new CodeExpansionInfo(
+                kernelSpec.Key,
+                CodeExpansion.CodeExpansionKind.KernelSpecConnection,
+                kernelSpec.Value.DisplayName);
+
+            var codeExpansion = new CodeExpansion(
+                [new($"#!connect jupyter --kernel-name {kernelSpec.Key} --kernel-spec {kernelSpec.Key}", "csharp")],
+                codeExpansionInfo);
+
+            _codeExpansions.Add(
+                codeExpansionInfo.Name,
+                codeExpansion);
+        }
+
+        _checkedKernelSpecs = true;
```

Comment:
For another iteration, we touched on this in our offline walkthrough of the PR, it would be nice to have a means of refreshing this cache based on a counter.

Reply:
Refreshing it in-process is less interesting than adding a cache's longer-lived than the process but that's for another day.